src/test/java/org/apache/commons/crypto/stream/AbstractCipherStreamTest.java

```diff
@@ -197,46 +224,236 @@ protected void doByteBufferRead(final String cipherClass, final boolean withChan
+        CryptoCipher cipher = getCipher(cipherClass);
+        String transformation = cipher.getAlgorithm();
+        out = getCryptoOutputStream(transformation, props, baos, key, 
+                new IvParameterSpec(iv), withChannel);
+        doByteBufferWrite(out, withChannel);
+        out.write(1);
+        Assert.assertTrue(out.isOpen()); 
+        
+        out = getCryptoOutputStream(transformation, props, baos, key, 
+                new IvParameterSpec(iv), withChannel);
+        out.close();
+        Assert.assertTrue(!out.isOpen());
+    }
 
-        Assert.assertEquals(dataLen, n1 + n2 + n3);
+    protected void doExceptionTest(final String cipherClass, ByteArrayOutputStream baos,
+            final boolean withChannel) throws IOException {
+        if (AbstractCipherTest.OPENSSL_CIPHER_CLASSNAME.equals(cipherClass)) {
+            if (!Crypto.isNativeCodeLoaded()) {
+                return; // Skip this test if no JNI
+            }
+        }
+        
+        InputStream in = null;
+        OutputStream out = null;
+        try { // Test InvalidAlgorithmParameters
+        	in = getCryptoInputStream(transformation, props, new ByteArrayInputStream(encData), 
+                    new SecretKeySpec(key, "AES"), new GCMParameterSpec(0, new byte[0]), 
+                    withChannel);
+            Assert.fail("Expected IOException.");
+        } catch (IOException ex) {
+            Assert.assertEquals(ex.getMessage(),"Illegal parameters");
+        } 
+        
+        try { // Test InvalidAlgorithmParameters
+            out = getCryptoOutputStream(transformation, props, baos, 
+                    new SecretKeySpec(key, "AES"), new GCMParameterSpec(0, 
+                    new byte[0]), withChannel);
+            Assert.fail("Expected IOException.");
+        } catch (IOException ex) {
+        	Assert.assertEquals(ex.getMessage(),"Illegal parameters");
+        } 
+        
+        try { // Test Invalid Key
+            in = getCryptoInputStream(transformation,props, new ByteArrayInputStream(encData), 
+                    new SecretKeySpec(new byte[10], "AES"), new IvParameterSpec(iv), withChannel);
+            Assert.fail("Expected IOException for Invalid Key");
+        } catch (IOException ex) {
+            Assert.assertNotNull(ex);
+        }
+        
+        try { // Test Invalid Key
+            out = getCryptoOutputStream(transformation, props, baos, new byte[10], 
+                    new IvParameterSpec(iv), withChannel);
+            Assert.fail("Expected IOException for Invalid Key");
+        } catch (IOException ex) {
+            Assert.assertNotNull(ex);
+        }
+        
+        try { // Test closed stream.
+            in = getCryptoInputStream(new ByteArrayInputStream(encData), 
+                    getCipher(cipherClass), defaultBufferSize, iv, withChannel);
+            in.close();
+            in.close(); // Don't throw exception.
+            in.read(); // Throw exception.
```

Comment:
Why 2 calls to close()?

Reply:
I needed to close a closed stream to hit the return branch and cover the line.  
  public void close() throws IOException {
        if (closed) {
            return;
        }

Reply:
I see. Then should we split this test into 2 try blocks?  If someone removes that path you're testing with the 2nd close, the test will still pass because you'd catch an IOException from the 2nd close failing instead of from the read() call.

To be honest, I'd expect a double close() to throw and exception, but obviously not in this case.

Reply:
Interesting point.  I did some digging.  In that scenario, if an IOException were thrown at the second call to close, the test would fail at line 393 unless the exception thrown happened to have the same message.  

Nevertheless, I ran the tests with the branch in question removed and the test passes because the second close doesn't throw an exception at all (I stepped through it).  The method contract forces us to deal with an IOException, but it turns out that one isn't thrown in the existing implementation.

Reply:
Can you still split the test into 2?  One to test double close(), and one to test operating on a closed stream?  That way it is explicit there are two cases being tested here.

Reply:
LGTM